Readme.md

```diff
@@ -371,16 +371,34 @@ Optionally, a second parameter can be given. The remaining free size of the area
 
 ### Auto areas
 
-To help manage allocating new data in existing space, you can use `.autoarea` for armips to automatically find an area with enough space.  Example:
+To help manage allocating new data in existing space, you can use `.sharedarea` and `.autoarea` for armips to automatically find an area with enough space.
+
+`.sharedarea` uses the same parameters as `.area`, but marks the area as shared for `.autoarea` to use.
+
+Example `.autoarea` usage:
 
 ```
+.org @FreeSpace
+.sharedarea 0x4000
+.byte 0x12, 0x34
+.endsharedarea
+
 .autoarea
 @TheAnswer:
   .byte 42
 .endautoarea
 ```
```

Comment:
You're talking about this scenario, specifically, right?

```
.definesharedarea 0x082345678,0x1000,0x00

.autoarea
.func MyFunc
bx 14
.endfunc
.endautoarea

; Keep the party going.
dw 0x1337C0DE
```

In this scenario, it would keep writing after the end of the automatically allocated content, and there would be no error.  Specifically, it would write 0x1337C0DE to 0x08234567A, which might even be overwritten by another later `.autoarea`.

In theory it could move you to the end of that sharedarea (this is probably never useful) or perhaps restore the old position (possibly useful?)

Not sure if there's a great mechanism to mark the current address "invalid" and report as an error if you don't use org before assembly more.  That's probably the best idea, and least surprising.

The main purpose of `.area` in general is to keep yourself constrained, so I'd argue that this is a general problem.  It might be more obvious where you'll be after `.endarea`, but it's just as likely "wrong" to keep writing then as well.  Might see about adding this as a general error for all `.area` variants as a separate pull request.

-[Unknown]

Reply:
Writing more code after an `.area` is not an error IMO, because `.area` has a fixed size, therefore you should be at a predictable address after it ends.

IIRC this is how it works currently:
```
.org 0x2123456
.area 0x30
    // a bunch of code
    b       @end
.endarea
@end: // 0x2123486
    pop     r15
```

Reply:
Yep, exactly that scenario. Couple points there...

1. That isn't how `.area` currently works; it defines a limitation, rather than enforcing that all that space must be taken up. The address you end up at is thus not predictable without evaluating what is within the `.area`. Unless, of course, you specify a byte with which to fill the unused space. (Does it fill this space before, or after, any `.autoarea`s consume the unused space? Ought to be before, I would think.)
2. `.autoarea` is a completely different kind of operation than `.area` or `.sharedarea`. It relocates the contained code somewhere else, which IIRC isn't something that anything else in armips previously did.
3. It's probably "wrong" to assemble after an `.area` without a fill byte specified, but not especially surprising. It's perfectly valid to assemble after an `.area` with a fill byte specified. It's definitely surprising, and bad behavior, if code/data assembled after an `.autoarea` is overwritten by a later `.autoarea`, or if it doesn't respect the bounds of the `.sharedarea`.

Given this and @unknownbrackets 's commentary, the behavior should probably be either to restore the old position, or throw a warning/error.

Reply:
For a `.sharedarea`, the fill parameter is applied after any allocation.  Consider:

```
.definesharedarea 0x08234567,0x1000,0x00
```

Any allocations into this shared area will be made first at 0x08234567, decreasing the remaining space to fill.  The filled bytes will be after.

As far as the order innternally within armips, it actually goes like this:

1. Validation pass 1: Area direct content and fill are in the area (same as before this change.)  Allocations are not placed anywhere yet.
2. Validation pass 2: Area direct content + allocations are put in the shared areas, and fill amount decreases (staying at end.)
3. Validation pass N: If moving the content into the shared area caused other things to recalculate (labels, etc.) they are re-evaluated until they stop changing (as per normal armips design.)

I've changed the behavior to reset to the previous position, but I still think any warnings/errors should be a separate change.  Although there are valid use cases (such as structs with zero fill) for using `.area` with fill, I think any warnings without fill/etc. should be considered more generally because they are not only relevant to any allocated area.

As far as the name - I'm not really tied to `.autoarea`, although I think it should be clear that it fits into areas.  I think most people will want to use it the same way I want to use it:

1. I'm currently using `.area` to replace existing functions in the game with functions friendly to localization.
2. In virtually every case, my implementation is more efficient (the compiler made some really terrible choices), so even after adding new features, VWF, centering logic, etc. 33% of the bytes I've replaced is still free.
3. The game has some ugly hacks, like printing "   's friends can be seen here" and then drawing the main character's name on top.  I replaced this with a simple control code i.e. "See stats for [NAME1] and friends."
4. The control code needed a function and didn't fit right in the text rendering loop (I'd already added so much there), but I knew I had several kilobytes of free space nearby.  I had to manually calculate where was a good spot, and I'd rather have software do the busy work for me.
5. I also didn't like having to put labels in obscure places to "fit" code in there, when it didn't semantically belong to that area (in the code, who cares about the output as long as it's a reachable area.)

So in my mind, it's still very conceptually linked to areas.  To me, they are like whiteout for the game's original data.  If I was specifically trying to fill a section with 0x00, I'd use `.fill`, because that conceptually makes more sense.

If we rename it, `.sharedarea` should probably be renamed to be consistent.

-[Unknown]